File: js/test/Ice/operations/Oneways.js
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

const Ice = require("ice").Ice;
const test = require("TestHelper").TestHelper.test;

async function run(communicator, prx, Test, bidir)
{
    prx = prx.ice_oneway();
    await prx.ice_ping();

    try
    {
        await prx.ice_isA(Test.MyClass.ice_staticId());
        test(false);
    }
    catch(ex)
    {
        // Expected: twoway proxy required
        test(ex instanceof Ice.TwowayOnlyException, ex);
    }

    try
    {
        await prx.ice_id();
        test(false);
    }
    catch(ex)
    {
        // Expected: twoway proxy required
        test(ex instanceof Ice.TwowayOnlyException, ex);
    }

    try
    {
        await prx.ice_ids();
        test(false);
    }
    catch(ex)
    {
        // Expected: twoway proxy required
        test(ex instanceof Ice.TwowayOnlyException, ex);
    }

    await prx.opVoid();
    await prx.opIdempotent();

    try
    {
        await prx.opByte(0xff, 0x0f);
        test(false);
    }
    catch(ex)
    {
        // Expected: twoway proxy required
        test(ex instanceof Ice.TwowayOnlyException, ex);
    }
}

exports.Oneways = {run: run};

File: cpp/test/Ice/operations/Oneways.cpp
// Copyright (c) ZeroC, Inc.

#include "Ice/Ice.h"
#include "Test.h"
#include "TestHelper.h"

using namespace std;

void
oneways(const Ice::CommunicatorPtr&, const Test::MyClassPrx& proxy)
{
    Test::MyClassPrx p = proxy->ice_oneway();

    {
        p->ice_ping();
    }

    {
        try
        {
            [[maybe_unused]] bool _ = p->ice_isA("dummy");
            test(false);
        }
        catch (const Ice::TwowayOnlyException&)
        {
        }
    }

    {
        try
        {
            [[maybe_unused]] string _ = p->ice_id();
            test(false);
        }
        catch (const Ice::TwowayOnlyException&)
        {
        }
    }

    {
        try
        {
            [[maybe_unused]] vector<string> _ = p->ice_ids();
            test(false);
        }
        catch (const Ice::TwowayOnlyException&)
        {
        }
    }

    {
        p->opVoid();
    }

    {
        p->opIdempotent();
    }

    {
        uint8_t b;

        try
        {
            p->opByte(uint8_t(0xff), uint8_t(0x0f), b);
            test(false);
        }
        catch (const Ice::TwowayOnlyException&)
        {
        }
    }
}
